cmd/root_test.go

```diff
@@ -19,157 +19,196 @@ package cmd
+		subDir := filepath.Join(path, "subidr")
+		if err = os.MkdirAll(subDir, os.ModePerm); err != nil {
+			c.Fatal(err)
+		}
+		if err = os.Chdir(subDir); err != nil {
+			c.Fatal(err)
+		}
+		r, dir, err := hpcToolkitRepo()
+		c.Assert(err, IsNil)
+		c.Check(dir, Equals, subDir)
+		h, _ := r.Head()
+		c.Check(h.Hash(), Equals, head.Hash())
+	}
+
+	{ // CWD is root of sub repo in repo root
+		subRepo := filepath.Join(path, "subrepo")
+		if err = os.MkdirAll(subRepo, os.ModePerm); err != nil {
+			c.Fatal(err)
+		}
+		_, _, err = initTestRepo(subRepo)
+		if err != nil {
+			c.Fatal(err)
+		}
+		if err = os.Chdir(subRepo); err != nil {
+			c.Fatal(err)
+		}
+		r, dir, err := hpcToolkitRepo()
+		c.Assert(err, IsNil)
+		c.Check(dir, Equals, path)
+		h, _ := r.Head()
+		c.Check(h.Hash(), Equals, head.Hash())
+	}
+
+	{ // CWD is parent of repo root, hope it's not repo itself.
+		os.Chdir(filepath.Dir(path))
+		_, _, err = hpcToolkitRepo()
+		c.Check(err, Equals, git.ErrRepositoryNotExists)
+	}
 }
 
 func (s *MySuite) TestIsHpcToolkitRepo(c *C) {
-	// sub directory of an hpc-toolkit git repository
-	_, callDir, _, _ := runtime.Caller(0)
-	repoDir := filepath.Dir(filepath.Dir(callDir))
-	repo, _ := git.PlainOpen(repoDir)
-	got := isHpcToolkitRepo(*repo)
-	c.Assert(got, Equals, true)
-
-	// temporary git repo is not a hpc-toolkit repo
-	storer := memory.NewStorage()
-	fs := memfs.New()
-	testRepo, _ := git.Init(storer, fs)
-	initTestRepo(*testRepo, fs)
-	got = isHpcToolkitRepo(*testRepo)
-	c.Assert(got, Equals, false)
+	repo, initHash, err := initTestRepo(c.MkDir())
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	// Doesn't match
+	GitInitialHash = randomGitHash
+	c.Check(isHpcToolkitRepo(*repo), Equals, false)
+
+	// Matches
+	GitInitialHash = initHash.String()
+	c.Check(isHpcToolkitRepo(*repo), Equals, true)
 }
 
 func (s *MySuite) TestCheckGitHashMismatch(c *C) {
-	_, callDir, _, _ := runtime.Caller(0)
-	repoDir := filepath.Dir(filepath.Dir(callDir))
-	workDir := filepath.Dir(callDir)
-	repo, _ := git.PlainOpen(repoDir)
+	path := c.MkDir()
+	repo, init, err := initTestRepo(path)
+	if err != nil {
+		c.Fatal(err)
+	}
+	GitInitialHash = init.String()
+	if err = os.Chdir(path); err != nil {
+		c.Fatal(err)
+	}
+
 	head, _ := repo.Head()
 	hash := head.Hash().String()
 	branch := head.Name().Short()
 
-	// verify current working directory git hash against hash of v1.0.0
-	mockInjectedGitVariables()
-	mismatch, b, h, dir := checkGitHashMismatch()
-	c.Assert(mismatch, Equals, true)
-	c.Assert(dir, Equals, workDir)
-	c.Assert(b, Equals, branch)
-	c.Assert(h, Equals, hash)
+	{ // Matches
+		GitCommitHash = hash
+		mismatch, _, _, _ := checkGitHashMismatch()
+		c.Check(mismatch, Equals, false)
+	}
 
-	// verify current working directory git hash against random initial hash
-	mockInjectedGitVariables()
-	GitInitialHash = randomGitHash
-	mismatch, b, h, dir = checkGitHashMismatch()
-	c.Assert(mismatch, Equals, false)
-	c.Assert(b, Equals, "")
-	c.Assert(h, Equals, "")
```

Comment:
nit: Before we verified that in the non-mismatch case we don't emit any spurious data. There is nothing binding us to that contract but I kind of like it in these named return functions which make it easy for random data to slip out.

Reply:
I see this modality several places, I could be convinced otherwise. Thoughts?